### PR TITLE
Add tests for deepcopy of structs

### PR DESCRIPTION
**Pull request by liggitt**
Until https://github.com/kubernetes/kubernetes/pull/35728 merges, we want to at least fuzz/test that deepcopy isn't shallow-copying problematic fields

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35912)

### CODE REVIEW

**Comment by liggitt** [comment]
#35898 flake
@k8s-bot unit test this

**Comment by kargakis** [comment]
> #35898 flake

I assume this was older than the fix (https://github.com/kubernetes/kubernetes/pull/36003) for the flake

**Comment by erictune** [comment]
@caesarxuchao can you review or reassign please.

File: pkg/api/copy_test.go [hunk 1]
```diff
@@ -57,6 +60,53 @@ func doDeepCopyTest(t *testing.T, kind unversioned.GroupVersionKind, f *fuzz.Fuz
...
+
+	// Most runtime.Objects are pointers
+	// Make sure calling DeepCopy() with a non-pointer struct either errors, or actually deep copies
+	itemValue := reflect.ValueOf(item)
+	if itemValue.Kind() == reflect.Ptr {
+		itemStructValue := itemValue.Elem()
+		fmt.Println(itemStructValue.Type())
+		itemStruct := itemStructValue.Interface()
+		itemStructCopy, err := api.Scheme.DeepCopy(itemStruct)
+		if err == nil || strings.Contains(err.Error(), "must pass addressable value") {
```
Comment (caesarxuchao) [inline pkg/api/copy_test.go]: `If err!=nil && strings.Contains(...`

File: pkg/api/copy_test.go [hunk 2]
```diff
@@ -57,6 +60,53 @@ func doDeepCopyTest(t *testing.T, kind unversioned.GroupVersionKind, f *fuzz.Fuz
...
+			t.Errorf("Could not deep copy struct of %v: %s", kind, err)
+			return
+		}
+		if !reflect.DeepEqual(itemStruct, itemStructCopy) {
+			t.Errorf("\nexpected: %#v\n\ngot:      %#v\n\ndiff:      %v", itemStruct, itemStructCopy, diff.ObjectReflectDiff(itemStruct, itemStructCopy))
+			return
+		}
+		// Refuzz the copy, which should have no effect on the original
+		f.Fuzz(itemStructCopy)
+		if reflect.DeepEqual(itemStruct, itemStructCopy) {
```
Comment (caesarxuchao) [inline pkg/api/copy_test.go]: I thought we should compare `itemStructPreFuzz` with `itemStructPostFuzz` here.

**Review by caesarxuchao — CHANGES_REQUESTED** [review]

**Comment by liggitt** [comment]
removed the struct copy stanza, still working on that, the rest is ready

**Comment by liggitt** [comment]
comments addressed, PTAL

**Comment by caesarxuchao** [comment]
LGTM. Thanks!

**Comment by saad-ali** [comment]
Release-czar approved post-code freeze merge--This was LGTMed and in the merge-queue at code freeze time for 1.5. Adding 1.5 milestone to let it gets merged after code freeze.

**Pull request merged**